README.md
# DEEL BACKEND TASK

## 👀 README INSIDE README 👀

First of all, thank you for this opportunity. I had a good time solving this exciting exercise. I ran out of time and couldn't finish it as I'd wished (my bad, I started yesterday morning, as you will see on my git logs).
However, I created a CHANGELOG.md file that summarizes what I did and would have done.
I also left a couple of comments inside the code in areas where I'd search for help if this were a real live exercise.
Lastly, don't forget to visit the [(uncompleted) Postman Documentation](https://documenter.getpostman.com/view/7063210/2s8Z6u5avF) that I created to test my work and make yours easy.

Enjoy!

## 💫 Welcome! 🎉

This backend exercise involves building a Node.js/Express.js app that will serve a REST API. We imagine you should spend around 3 hours at implement this feature.

## Data Models

> **All models are defined in src/model.js**

### Profile

A profile can be either a `client` or a `contractor`.
clients create contracts with contractors. contractor does jobs for clients and get paid.
Each profile has a balance property.

### Contract

A contract between a client and a contractor.
Contracts have 3 statuses, `new`, `in_progress`, `terminated`. contracts are considered active only when in status `in_progress`
Contracts group jobs within them.

### Job

contractor get paid for jobs by clients under a certain contract.

## Getting Set Up

The exercise requires [Node.js](https://nodejs.org/en/) to be installed. We recommend using the LTS version.

1. Start by cloning this repository.

1. In the repo root directory, run `npm install` to gather all dependencies.

1. Next, `npm run seed` will seed the local SQLite database. **Warning: This will drop the database if it exists**. The database lives in a local file `database.sqlite3`.

1. Then run `npm start` which should start both the server and the React client.

❗️ **Make sure you commit all changes to the master branch!**

## Technical Notes

- The server is running with [nodemon](https://nodemon.io/) which will automatically restart for you when you modify and save a file.

- The database provider is SQLite, which will store data in a file local to your repository called `database.sqlite3`. The ORM [Sequelize](http://docs.sequelizejs.com/) is on top of it. You should only have to interact with Sequelize - **please spend some time reading sequelize documentation before starting the exercise.**

- To authenticate users use the `getProfile` middleware that is located under src/middleware/getProfile.js. users are authenticated by passing `profile_id` in the request header. after a user is authenticated his profile will be available under `req.profile`. make sure only users that are on the contract can access their contracts.
- The server is running on port 3001.

## APIs To Implement

Below is a list of the required API's for the application.

1. **_GET_** `/contracts/:id` - This API is broken 😵! it should return the contract only if it belongs to the profile calling. better fix that!
2. **_GET_** `/contracts` - Returns a list of contracts belonging to a user (client or contractor), the list should only contain non terminated contracts.
3. **_GET_** `/jobs/unpaid` - Get all unpaid jobs for a user (**_either_** a client or contractor), for **_active contracts only_**.
4. **_POST_** `/jobs/:job_id/pay` - Pay for a job, a client can only pay if his balance >= the amount to pay. The amount should be moved from the client's balance to the contractor balance.
5. **_POST_** `/balances/deposit/:userId` - Deposits money into the balance of a Client, a Client can't deposit more than 25% his total of jobs to pay (at the deposit moment).
6. **_GET_** `/admin/best-profession?start=<date>&end=<date>` - Returns the profession that earned the most money (sum of jobs paid) for any Contractor that worked in the query time range.
7. **_GET_** `/admin/best-clients?start=<date>&end=<date>&limit=<integer>` - Returns the Clients that paid the most for jobs in the query time period. limit query parameter should be applied, default limit is 2.

```
 [
    {
        "id": 1,
        "fullName": "Reece Moyer",
        "paid" : 100.3
    },
    {
        "id": 200,
        "fullName": "Debora Martin",
        "paid" : 99
    },
    {
        "id": 22,
        "fullName": "Debora Martin",
        "paid" : 21
    }
]
```

## Going Above and Beyond the Requirements

Given the time expectations of this exercise, we don't expect anyone to submit anything super fancy, but if you find yourself with extra time, any extra credit item(s) that showcase your unique strengths would be awesome! 🙌

It would be great for example if you'd write some unit test / simple frontend demonstrating calls to your fresh APIs.

## Submitting the Assignment

When you have finished the assignment, create a github repository and send us the link.

Thank you and good luck! 🙏


CHANGELOG.md
### Version 0.2.0

This version includes possible breaking changes. Read the log carefully.

14. TODO: Finish API documentation. I started using POSTMAN. [Here you can find it](https://documenter.getpostman.com/view/7063210/2s8Z6u5avF).
13. TODO: Add a logger. One logger saves million lives.
12. TODO: Add a complete error handling strategy to cover most of the errors and send proper API error messages. 
11. TODO: Refactorize folder structure and Project/APIs architecture. I'd use some clean architecture approach by separating code into layers. Something like: routes -> expressCallback(optional) -> middlewares(auth, profile) -> controllers -> useCases -> dataAccess -> response/error-middlewares.
10. TODO: Create unit/integration test to verify all the APIs functionality, probably starting with those that involve money transactions. I'd use jest and supertest libraries.
9. Create a new /api/v1/admin/best-clients?start=<date>&end=<date>&limit=<number> endpoint that returns the Clients that paid the most for jobs in the query time period. Limit query parameter should be applied, default limit is 2.
8. Create a new /api/v1/admin/best-profession?start=<date>&end=<date> endpoint that returns the profession that earned the most money (sum of jobs paid) for any Contractor that worked in the query time range.
7. Create a new /api/v1/balances/deposit/:userId endpoint to deposit money into the balance of a Client (userId). A Client can't deposit more than 25% his total of jobs price to pay (at the deposit moment).
6. Create a new /api/v1/jobs/:jobId/pay endpoint to POST a payment from a Client to a Contractor. A Client can only pay if his balance >= the amount to pay, and if the Contract is active. The amount should be moved from the Client's balance to the Contractor balance.
5. Create a new /api/v1/jobs/unpaid endpoint to GET All the unpaid jobs for a user (**_either_** a client or contractor), for **_active contracts only_**.
4. Create a new /api/v1/contracts endpoint to GET All the contracts only if they belong to the profile calling. Create a new ContractStatus enum to handle this kind of word (properties) and better avoid typos.
3. Fix models and database configuration that was causing the 'seed' script (sequelize) to make wrong table relationships. Fix the /api/v1/contracts/:contractId endpoint to GET the contract only if it belongs to the profile calling.
2. Migrate code from JavaScript to TypeScript. Separate models into single files and move the database connection to the config folder. Add a custom type to Express Response and Request interfaces to be allowed to use req.profile property. Remove the body-parser library to use the express.json() method that Express has included since version 4.
1. Install required dependencies and create config files to run the project using TypeScript, ESLint, Prettier, and Nodemon. Modify package.json to include some new scripts. Add a script to run in debugging mode. Add CHANGELOG file.


src/app.ts
import express from 'express';
import { Op } from 'sequelize';
import { sequelize } from './models';
import { getProfile } from './middleware/getProfile.middleware';
import { ContractStatus } from './types/contract.types';
const app = express();

app.use(express.json());
app.set('sequelize', sequelize);
app.set('models', sequelize.models);

/**
 * @description Get a contract if it is related to the profile
 * @returns a contract object or an error
 */
app.get('/api/v1/contracts/:contractId', getProfile, async (req, res) => {
  const { Contract } = req.app.get('models');
  const { contractId } = req.params;
  const { profile } = req;
  if (!profile) return res.status(404).json({ error: 'Contract not found' });

  const contract = await Contract.findOne({
    where: { id: contractId, [Op.or]: [{ clientId: profile.id }, { contractorId: profile.id }] },
  });
  if (!contract) return res.status(404).json({ error: 'Contract not found' });
  res.status(200).json({ data: { contract } });
});

/**
 * @description Get all contracts related to the profile
 * @returns an array of contracts or an error
 */
app.get('/api/v1/contracts', getProfile, async (req, res) => {
  const { Contract } = req.app.get('models');
  const { profile } = req;
  if (!profile) return res.status(404).json({ error: 'No Contracts found' });

  const contracts = await Contract.findAll({
    where: {
      [Op.or]: [{ clientId: profile.id }, { contractorId: profile.id }],
      status: { [Op.ne]: ContractStatus.TERMINATED },
    },
  });
  if (!contracts || !contracts.length) {
    return res.status(200).json({ message: 'There are no Contracts related to this Profile' });
  }

  res.status(200).json({ data: { contracts } });
});

/**
 * @description Get all unpaid jobs for a client or contractor for contracts that are in progress
 * @returns an array of jobs or an error
 */
app.get('/api/v1/jobs/unpaid', getProfile, async (req, res) => {
  const { Contract, Job } = req.app.get('models');
  const { profile } = req;
  if (!profile) return res.status(404).json({ error: 'No Jobs found' });

  const jobs = await Contract.findAll({
    attributes: [['id', 'contractId'], 'clientId', 'contractorId', 'status'],
    where: {
      [Op.or]: [{ clientId: profile.id }, { contractorId: profile.id }],
      status: ContractStatus.IN_PROGRESS,
    },
    include: {
      model: Job,
      as: 'jobs',
      where: { [Op.or]: [{ paid: false }, { paid: null }] },
    },
  });

  // This is another way to do it. It'd be nice to discover which is more performant.
  /* const jobs = await Job.findAll({
    where: {
      [Op.or]: [{ paid: paid === 'true' }, { paid: null }],
      contractId: {
        [Op.in]: sequelize.literal(
          `(SELECT id FROM Contracts WHERE clientId = ${profile.id} OR contractorId = ${profile.id})`,
        ),
      },
    },
  }); */

  if (!jobs || !jobs.length) {
    return res.status(200).json({ message: 'There are no Jobs related to this Profile' });
  }

  res.status(200).json({ data: { jobs } });
});

/**
 * @description Pay for a job which has a contract status in_progress. A client can only pay if his balance >= the amount to pay. The amount should be moved from the client's balance to the contractor balance.
 * @returns a success message or an error
 */
app.post('/api/v1/jobs/:jobId/pay', getProfile, async (req, res) => {
  const { Contract, Job, Profile } = req.app.get('models');
  const { jobId } = req.params;
  const { profile } = req;
  if (!profile) return res.status(404).json({ error: 'Job not found' });

  const job = await Job.findOne({
    where: { id: jobId, [Op.or]: [{ paid: false }, { paid: null }] },
    include: [
      {
        model: Contract,
        as: 'contract',
        where: { clientId: profile.id, status: ContractStatus.IN_PROGRESS },
      },
    ],
  });
  if (!job)
    return res.status(404).json({
      error: 'Job already paid, not found, or the Contract for this Profile is not active',
    });

  const { price } = job;
  if (profile.balance < price) {
    return res.status(400).json({ error: 'Insufficient funds' });
  }

  try {
    await sequelize.transaction(async (t) => {
      await Profile.update(
        { balance: sequelize.literal(`balance - ${price}`) },
        { where: { id: job.contract.clientId } },
        { transaction: t },
      );

      await Profile.update(
        { balance: sequelize.literal(`balance + ${price}`) },
        { where: { id: job.contract.contractorId } },
        { transaction: t },
      );

      await Job.update({ paid: true }, { where: { id: jobId } }, { transaction: t });
    });

    res.status(200).json({ message: 'Job paid successfully' });
  } catch (error) {
    res.status(500).json({ error: 'Something went wrong' });
  }
});

/**
 * @description Deposits money into the balance of a Client, a Client can't deposit more than 25% his total of jobs price to pay (at the deposit moment).
 * @returns a success message or an error
 */
app.post('/api/v1/balances/deposit/:userId', getProfile, async (req, res) => {
  const { Contract, Job, Profile } = req.app.get('models');
  const { userId } = req.params;
  const { profile } = req;

  if (!profile) return res.status(404).json({ error: 'Profile not found' });
  if (profile.id !== Number(userId)) {
    return res.status(403).json({ error: 'You can not deposit for another Profile' });
  }

  const { amount } = req.body;
  if (!amount || amount <= 0) {
    return res.status(400).json({ error: 'Amount is required and must be greater than 0' });
  }

  const { balance } = profile;
  const totalJobsToPay = await Job.sum('price', {
    where: { [Op.or]: [{ paid: false }, { paid: null }] },
    include: [
      {
        model: Contract,
        as: 'contract',
        where: { clientId: profile.id, status: ContractStatus.IN_PROGRESS },
      },
    ],
  });

  const maxDeposit = totalJobsToPay * 1.25;

  if (totalJobsToPay === 0) {
    return res.status(400).json({
      error: 'You can not deposit money because you have no jobs to pay',
    });
  }

  if (balance + amount > maxDeposit) {
    return res.status(400).json({
      error: `You can not deposit more than 25% of your total jobs to pay. Please pay some jobs first. MaxDeposit: ${maxDeposit}, Balance: ${balance}, MaxAmount: ${(
        maxDeposit - balance
      ).toFixed(2)}`,
    });
  }

  try {
    await Profile.update(
      {
        balance: sequelize.literal(`
        CASE WHEN balance IS NULL THEN ${amount} ELSE balance + ${amount} END
      `),
      },
      { where: { id: userId } },
    );

    return res.status(200).json({ message: 'Deposit successful' });
  } catch (error) {
    return res.status(500).json({ error: 'Something went wrong' });
  }
});

/**
 * @description Returns the profession that earned the most money (sum of jobs paid) for any Contractor that worked in the query time range.
 * @returns a success message or an error
 */
app.get('/api/v1/admin/best-profession', async (req, res) => {
  const { Contract, Job, Profile } = req.app.get('models');
  const { start, end } = req.query;

  if (!start || !end) {
    return res.status(400).json({ error: 'Start and End dates are required' });
  }

  const bestProfession = await Profile.findAll({
    attributes: [
      'profession',
      [sequelize.fn('sum', sequelize.col('contractor.jobs.price')), 'total'],
    ],
    where: {
      id: {
        [Op.in]: sequelize.literal(
          `(SELECT contractorId FROM Contracts WHERE id IN (SELECT contractId FROM Jobs WHERE paid = true AND paymentDate BETWEEN '${start}' AND '${end}'))`,
        ),
      },
    },
    include: [
      {
        model: Contract,
        as: 'contractor',
        where: {
          id: {
            [Op.in]: sequelize.literal(
              `(SELECT contractId FROM Jobs WHERE paid = true AND paymentDate BETWEEN '${start}' AND '${end}')`,
            ),
          },
        },
        include: [
          {
            model: Job,
            as: 'jobs',
            where: { paid: true, paymentDate: { [Op.between]: [start, end] } },
          },
        ],
      },
    ],
    group: ['profession'],
    order: [[sequelize.literal('total'), 'DESC']],
  });

  if (!bestProfession || !bestProfession.length) {
    return res.status(200).json({ message: 'There are no Jobs in this period' });
  }

  return res.status(200).json({ data: { bestProfession: bestProfession[0].profession } });
});

/**
 * @description Returns the Clients that paid the most for jobs in the query time period. limit query parameter should be applied, default limit is 2.
 * @example /api/v1/admin/best-clients?start=2020-01-01&end=2020-12-31&limit=3
 * @returns a success message or an error
 */
app.get('/api/v1/admin/best-clients', async (req, res) => {
  const { Contract, Job, Profile } = req.app.get('models');
  const { start, end, limit } = req.query;

  if (!start || !end) {
    return res.status(400).json({ error: 'Start and End dates are required' });
  }

  const bestClients = await Profile.findAll({
    // There is a bug in sequelize that does not allow to use limit when include is used. This is a workaround for that.
    limit: limit || 2,
    subQuery: false,
    // I found this solution in this issue: https://github.com/sequelize/sequelize/issues/9651#issuecomment-460160022
    attributes: [
      'id',
      'firstName',
      'lastName',
      [sequelize.fn('sum', sequelize.col('client.jobs.price')), 'total'],
    ],
    where: {
      id: {
        [Op.in]: sequelize.literal(
          `(SELECT clientId FROM Contracts WHERE id IN (SELECT contractId FROM Jobs WHERE paid = true AND paymentDate BETWEEN '${start}' AND '${end}'))`,
        ),
      },
    },
    include: [
      {
        model: Contract,
        as: 'client',
        where: {
          id: {
            [Op.in]: sequelize.literal(
              `(SELECT contractId FROM Jobs WHERE paid = true AND paymentDate BETWEEN '${start}' AND '${end}')`,
            ),
          },
        },
        required: true,
        include: [
          {
            model: Job,
            as: 'jobs',
            where: { paid: true, paymentDate: { [Op.between]: [start, end] } },
            required: true,
          },
        ],
      },
    ],
    group: ['firstName', 'lastName'],
    order: [[sequelize.literal('total'), 'DESC']],
  });

  if (!bestClients || !bestClients.length) {
    return res.status(200).json({ message: 'There are no Clients in this period' });
  }

  return res.status(200).json({ data: { bestClients } });
});

export default app;
